### PR TITLE
Add spec for IPFS gateway for Sticker Pack

### PR DESCRIPTION
**Pull request by Ferossgp**

### CODE REVIEW

File: docs/draft/sticker-pack.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+### Status sticker usage
+To send a sticker in chat, a user of Status should  buy or install a sticker pack.
+To install a sticker pack, we need to fetch all sticker packs which are available in Sticker Market. 
+The following steps are needed to fetch all sticker packs:
+#### 1. Get total number of sticker packs
+Call `packCount()` on the sticker market contract, will return number of sticker pack registered as `uint256`.
+#### 2. Get sticker pack by id
+ID's are represented as `uint256` and are incremental from `0` to totatl number of sticker packs in contract, which we received on previous step. To get a sticker pack we should call `getPackData(sticker-pack-id)`, the return type is  `["bytes4[]" "address" "bool" "uint256" "uint256" "bytes"]` which represents the following fields: `[category owner mintable timestamp price contenthash]`. 
+`TODO:` Describe contenthash, it's a hash in IPFS (what it contains?)
```
Comment (Ferossgp) [inline docs/draft/sticker-pack.md]: Contains pack info, describe the structure and all fields

File: docs/draft/sticker-pack.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+ 1. [Abstract](#abstract)
+ 2. [Specification](#specification)
+ 3. [Copyright](#copyright)
+
+## Abstract
+
+In this specification, we describe how Status app uses IPFS gateway to store stickers.
+We will explore image format, how they are uploaded and how an end user can see them
+insed the Status app.
```
Comment (andremedeiros) [inline docs/draft/sticker-pack.md]: insed -> inside

Comment (vkjr) [inline docs/draft/sticker-pack.md]: "insed" misprint

File: docs/draft/sticker-pack.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+### Distribution
+
+Sticker pack are implemented as ERC721 token and contains a set of stickers. These stickers
+are stored inside the sticker pack as a set of hyperlinks pointing to IPFS storage. These
+hyperlinks are publically available and can be accessed by any user inside the status chat.
+Stickers can be sent in chat only by accounts that own the sticker pack.
+
+### IPFS gateway
+At the moment of writing, the Status uses the [Infura](https://infura.io/) gateway.
```
Comment (andremedeiros) [inline docs/draft/sticker-pack.md]: the Status -> Status

File: docs/draft/sticker-pack.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+When a sticker is shown in the app, Status app makes an http GET request to IPFS gateway using the hyperlink. 
+
+### Status sticker usage
+To send a sticker in chat, a user of Status should  buy or install a sticker pack.
+To install a sticker pack, we need to fetch all sticker packs which are available in Sticker Market. 
+The following steps are needed to fetch all sticker packs:
+#### 1. Get total number of sticker packs
+Call `packCount()` on the sticker market contract, will return number of sticker pack registered as `uint256`.
+#### 2. Get sticker pack by id
+ID's are represented as `uint256` and are incremental from `0` to totatl number of sticker packs in contract, which we received on previous step. To get a sticker pack we should call `getPackData(sticker-pack-id)`, the return type is  `["bytes4[]" "address" "bool" "uint256" "uint256" "bytes"]` which represents the following fields: `[category owner mintable timestamp price contenthash]`.
```
Comment (yenda) [inline docs/draft/sticker-pack.md]:
```suggestion
ID's are represented as `uint256` and are incremental from `0` to total number of sticker packs in contract, which we received on previous step. To get a sticker pack we should call `getPackData(sticker-pack-id)`, the return type is  `["bytes4[]" "address" "bool" "uint256" "uint256" "bytes"]` which represents the following fields: `[category owner mintable timestamp price contenthash]`. 
```

File: docs/draft/sticker-pack.md (hunk 2 above)
Comment (corpetty) [inline docs/draft/sticker-pack.md:20]: ~insed~ inside

**Comment by andremedeiros** [comment]
Do we know whether the Infura gateway pins the IPFS content?

File: docs/draft/sticker-pack.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+#### 2. Get sticker pack by id
+ID's are represented as `uint256` and are incremental from `0` to totatl number of sticker packs in contract, which we received on previous step. To get a sticker pack we should call `getPackData(sticker-pack-id)`, the return type is  `["bytes4[]" "address" "bool" "uint256" "uint256" "bytes"]` which represents the following fields: `[category owner mintable timestamp price contenthash]`. 
+`TODO:` Describe contenthash, it's a hash in IPFS (what it contains?)
+##### 3. Get owned stick packs
+To get owned packs, we should get all owned tokens for the current account address. To do that we should call `balanceOf(address)` where address is the address for current account. This method returns a `uint256` representing the token id representing the count of available tokens. Using `tokenOfOwnerByIndex(address,uint256)` method, with the address of the user and ID in form of a `uint256` which is an incremented int from 0 to total number of tokens, we will get token id. To get sticker pack id from token we call`tokenPackId(uint256)` where `uint256` is the token id. This method will return an `uint256` which is the id of the owned sticker pack.
```
Comment (Ferossgp) [inline docs/draft/sticker-pack.md]:
```suggestion
To get owned packs, we should get all owned tokens for the current account address. To do that we should call `balanceOf(address)` where address is the address for current account. This method returns a `uint256` representing the count of available tokens. Using `tokenOfOwnerByIndex(address,uint256)` method, with the address of the user and ID in form of a `uint256` which is an incremented int from 0 to total number of tokens, we will get token id. To get sticker pack id from token we call`tokenPackId(uint256)` where `uint256` is the token id. This method will return an `uint256` which is the id of the owned sticker pack.
```

**Comment by hesterbruikman** [comment]
Is there a way to add when the method to `3. Get owned stick packs` is used? E.g. the instant account is recovered and address is known or upon opening, stickermarket.

File: docs/draft/sticker-pack.md (hunk 2 above)
Comment (Ferossgp) [inline docs/draft/sticker-pack.md]:
```suggestion
inside the Status app.
```

File: docs/draft/sticker-pack.md (hunk 3 above)
Comment (Ferossgp) [inline docs/draft/sticker-pack.md]:
```suggestion
At the moment of writing, Status uses the [Infura](https://infura.io/) gateway.
```

File: docs/draft/sticker-pack.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
+---
+title: IPFS gateway for Sticker Pack
+version: 0.1.0
+status: Draft
+authors: Gheorghe Pinzaru <gheorghe@status.im>
+redirect_from:
```
Comment (oskarth) [inline docs/draft/sticker-pack.md]: id missing

File: docs/draft/sticker-pack.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+## Specification
+
+### Image format
+Accepted image file types are **PNG, JPG/JPEG and GIF**, with a maximum allowed size of 300kb.
+The minimum sticker image resoulution is 512x512, and background of clipart should be transparent.
+
+### Distribution
+
+Sticker pack are implemented as ERC721 token and contains a set of stickers. These stickers
```
Comment (oskarth) [inline docs/draft/sticker-pack.md]: Link to ERC721 spec

File: docs/draft/sticker-pack.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+are stored inside the sticker pack as a set of hyperlinks pointing to IPFS storage. These
+hyperlinks are publically available and can be accessed by any user inside the status chat.
+Stickers can be sent in chat only by accounts that own the sticker pack.
+
+### IPFS gateway
+At the moment of writing, Status uses the [Infura](https://infura.io/) gateway.
```
Comment (oskarth) [inline docs/draft/sticker-pack.md]: Rephrase as "current main Status app" is using Infura. And something like "However clients could choose to run IPFS node..."~

File: docs/draft/sticker-pack.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+of security of the provider, meta information of its users can be leaked. Or if the
+provider servers are unavailable the access to the whole IPFS network will be lost.
+
+When a sticker is shown in the app, Status app makes an http GET request to IPFS gateway using the hyperlink. 
+
+### Status sticker usage
+To send a sticker in chat, a user of Status should  buy or install a sticker pack.
+To install a sticker pack, we need to fetch all sticker packs which are available in Sticker Market. 
+The following steps are needed to fetch all sticker packs:
+#### 1. Get total number of sticker packs
```
Comment (oskarth) [inline docs/draft/sticker-pack.md:86]: spacing

File: docs/draft/sticker-pack.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+At the moment of writing, Status uses the [Infura](https://infura.io/) gateway.
+Infura gateway is an HTTPS gateway, which based on an HTTP GET request with the
+multihash block will return the stored content at that block address. 
+
+The use of gateway is required to enable easy access to the resources over HTTP.
+Each image of a sticker is stored inside IPFS using a unique address that is 
+derived from the hash of the file. This ensures that a file can't be overridden,
+and an end-user of the IPFS will receive the same file at a given address.
+
+The IPFS gateway acts as an end-user of the IPFS and allows users of the gateway
```
Comment (oskarth) [inline docs/draft/sticker-pack.md]: Move to security consideration section, also what meta information? Seems like typo in "whole IPFS network will be lost".

Probably worth mentioning pinning requirement here too, though that's mgiht be more of a functional req/recommendation than security consideration

File: docs/draft/sticker-pack.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+to access IPFS without connection to the P2P network. Usage of a gateway introduces
+potential risk for the users of that gateway provider. In case of a compromising
+of security of the provider, meta information of its users can be leaked. Or if the
+provider servers are unavailable the access to the whole IPFS network will be lost.
+
+When a sticker is shown in the app, Status app makes an http GET request to IPFS gateway using the hyperlink. 
+
+### Status sticker usage
```
Comment (oskarth) [inline docs/draft/sticker-pack.md]: How are they uploaded by a content creator?

File: docs/draft/sticker-pack.md [hunk 12]
```diff
@@ -0,0 +1,77 @@
+---
+title: IPFS gateway for Sticker Pack
```
Comment (oskarth) [inline docs/draft/sticker-pack.md]: Should the focus of this spec be more on

1) How stickers are communicated, maintained, uploaded here, or
2) How IPFS is used for querying and posting data?

E.g. do we imagine this evolving into more elaborate sticker markets/images/videos etc, or towards IPFS light node/Swarm/Dagger usage?

File: docs/draft/sticker-pack.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+When a sticker is shown in the app, Status app makes an http GET request to IPFS gateway using the hyperlink. 
+
+### Status sticker usage
+To send a sticker in chat, a user of Status should  buy or install a sticker pack.
+To install a sticker pack, we need to fetch all sticker packs which are available in Sticker Market. 
+The following steps are needed to fetch all sticker packs:
+#### 1. Get total number of sticker packs
+Call `packCount()` on the sticker market contract, will return number of sticker pack registered as `uint256`.
+#### 2. Get sticker pack by id
+ID's are represented as `uint256` and are incremental from `0` to total number of sticker packs in contract, which we received on previous step. To get a sticker pack we should call `getPackData(sticker-pack-id)`, the return type is  `["bytes4[]" "address" "bool" "uint256" "uint256" "bytes"]` which represents the following fields: `[category owner mintable timestamp price contenthash]`.
```
Comment (oskarth) [inline docs/draft/sticker-pack.md]: What is price exactly?

File: docs/draft/sticker-pack.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+ 1. [Abstract](#abstract)
+ 2. [Specification](#specification)
+ 3. [Copyright](#copyright)
+
+## Abstract
+
+In this specification, we describe how Status app uses IPFS gateway to store stickers.
+We will explore image format, how they are uploaded and how an end user can see them
+inside the Status app.
```
Comment (oskarth) [inline docs/draft/sticker-pack.md]: Does it make sense to refer to PAYLOADS spec somewhere for how stickers are sent?

File: docs/draft/sticker-pack.md (hunk 11 above)
Comment (oskarth) [inline docs/draft/sticker-pack.md]: For the contract, are there any moderation / approval / control etc concerns? Or are they vanilla ERC721 tokens?

**Review by oskarth — CHANGES_REQUESTED** [review]
Good stuff! Some comments but looking quite solid

**Comment by Ferossgp** [comment]
@andremedeiros I could not find information if the Infura pins the content by default, but they have a method on the gateway to do that https://infura.io/docs/ipfs/get/pin-add

File: docs/draft/sticker-pack.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+## Table of Contents
+
+ 1. [Abstract](#abstract)
+ 2. [Specification](#specification)
+ 3. [Copyright](#copyright)
+
+## Abstract
+
+In this specification, we describe how Status app uses IPFS gateway to store stickers.
```
Comment (decanus) [inline docs/draft/sticker-pack.md]:
```suggestion
In this specification, we describe how Status uses the IPFS gateway to store stickers.
```

File: docs/draft/sticker-pack.md [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+| Term             | Description                                                                            |
+|------------------|----------------------------------------------------------------------------------------|
+| **Stickers**     | A set of images which can be used to express emotions                                  |
+| **Sticker Pack** | ERC721 token which includes the set of stickers                                        |
+| **IPFS**         | P2P network used to store and share data, in this case, the images for the stickerpack |
+
+## Specification
+
+### Image format
+Accepted image file types are **PNG, JPG/JPEG and GIF**, with a maximum allowed size of 300kb.
```
Comment (decanus) [inline docs/draft/sticker-pack.md]:
```suggestion
Accepted image file types are `PNG`, `JPG/JPEG` and `GIF`, with a maximum allowed size of 300kb.
```

File: docs/draft/sticker-pack.md [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+The minimum sticker image resoulution is 512x512, and background of clipart should be transparent.
```
Comment (decanus) [inline docs/draft/sticker-pack.md]:
```suggestion
The minimum sticker image resolution is 512x512, and its background SHOULD be transparent.
```

File: docs/draft/sticker-pack.md [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+### Distribution
+
+Sticker pack are implemented as [ERC721 token](https://eips.ethereum.org/EIPS/eip-721) and contains a set of stickers. These stickers
```
Comment (decanus) [inline docs/draft/sticker-pack.md]:
```suggestion
Sticker packs are implemented as [ERC721 token](https://eips.ethereum.org/EIPS/eip-721) and contain a set of stickers. These stickers
```

**Review by decanus — CHANGES_REQUESTED** [review]

**Comment by decanus** [comment]
One thing I am not understanding is why do we store a full URL, rather than the remote ID for a sticker? It seems counter intuitive to store the full URL.

File: docs/draft/sticker-pack.md (hunk 12 above)
Comment (Ferossgp) [reply docs/draft/sticker-pack.md]: This spec is about `2. How IPFS is used for querying and posting data`. Added link to Sticker Market spec where I think more detailed information is needed. Covered steps that are not directly connected with IPFS on a high level only.

**Comment by flexsurfer** [comment]
@decanus good question,  because as you can see here, https://ipfs.infura.io/ipfs/Qmcehb2dboemjLXDqLZ2V6Vf8w1Z1MVje9QSErERCMMUYs , its a latest stickerpack, we use contenthashes (https://eips.ethereum.org/EIPS/eip-1577), not URLs, we need to change this in spec

**Comment by decanus** [comment]
@flexsurfer glad we use hashes and not urls, please change in the spec @Ferossgp

File: docs/draft/sticker-pack.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+       :stickers [{:hash "e301017012207737b75367b8068e5bdd027d7b71a25138c83e155d1f0c9bc5c48ff158724495"}
+       {:hash "e301017012201a9cdea03f27cda1aede7315f79579e160c7b2b6a2eb51a66e47a96f47fe5284"}]}}
+```
+All assets fileds, are contenthash fileds as per [EIP 1577](https://eips.ethereum.org/EIPS/eip-1577).
+ This payload is uploaded also to IPFS, and the IPFS address is used in the content field of the Sticker Market contract. See [Sticker Market spec](https://github.com/status-im/sticker-market/blob/651e88e5f38c690e57ecaad47f46b9641b8b1e27/docs/specification.md) for a detailed description of the contract.
+
+#### Install a sticker pack
+
+To install a sticker pack, we need to fetch all sticker packs which are available in Sticker Market. The following steps are needed to fetch all sticker packs:
+#### 1. Get total number of sticker packs
```
Comment (oskarth) [inline docs/draft/sticker-pack.md:86]:
```suggestion

#### 1. Get total number of sticker packs
```

File: docs/draft/sticker-pack.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+Call `packCount()` on the sticker market contract, will return number of sticker pack registered as `uint256`.
+#### 2. Get sticker pack by id
```
Comment (oskarth) [inline docs/draft/sticker-pack.md:89]:
```suggestion

#### 2. Get sticker pack by id
```

File: docs/draft/sticker-pack.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+ID's are represented as `uint256` and are incremental from `0` to total number of sticker packs in contract, which we received on previous step. To get a sticker pack we should call `getPackData(sticker-pack-id)`, the return type is  `["bytes4[]" "address" "bool" "uint256" "uint256" "bytes"]` which represents the following fields: `[category owner mintable timestamp price contenthash]`. Price is the SNT value in wei setted by sticker pack owner. The contenthash is the IPFS address described in the [submit desction](#submit-a-sticker-pack) above. Other fields specification could be found in [Sticker Market spec](https://github.com/status-im/sticker-market/blob/651e88e5f38c690e57ecaad47f46b9641b8b1e27/docs/specification.md)
+##### 3. Get owned sticker packs
```
Comment (oskarth) [inline docs/draft/sticker-pack.md:92]:
```suggestion

##### 3. Get owned sticker packs
```

File: docs/draft/sticker-pack.md [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+The current Status app fetches owned sticker packs during the open of any sticker view (a screen which shows a sticker pack or the list of sticker packs).
+To get owned packs, we should get all owned tokens for the current account address. To do that we should call `balanceOf(address)` where address is the address for current account. This method returns a `uint256` representing the count of available tokens. Using `tokenOfOwnerByIndex(address,uint256)` method, with the address of the user and ID in form of a `uint256` which is an incremented int from 0 to total number of tokens, we will get token id. To get sticker pack id from token we call`tokenPackId(uint256)` where `uint256` is the token id. This method will return an `uint256` which is the id of the owned sticker pack.
+
+##### 4. Buy a sticker pack
+To buy a sticker pack we should call `approveAndCall(address,uint256,bytes)` where `address` is the address of buyer,`uint256` is the price and third parameters `bytes` is the callback  called if approved. In callback we call `buyToken(uint256,address,uint256)`, first parameter is sticker pack id, second buyers address, and the last is the price.
+## Copyright
```
Comment (oskarth) [inline docs/draft/sticker-pack.md:99]:
```suggestion

## Copyright
```

**Review by oskarth — APPROVED** [review]
Generally LGTM, some minor outstanding comments

File: docs/draft/sticker-pack.md [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+
+#### Install a sticker pack
+
+To install a sticker pack, we need to fetch all sticker packs which are available in Sticker Market. The following steps are needed to fetch all sticker packs:
+#### 1. Get total number of sticker packs
+Call `packCount()` on the sticker market contract, will return number of sticker pack registered as `uint256`.
+#### 2. Get sticker pack by id
+ID's are represented as `uint256` and are incremental from `0` to total number of sticker packs in contract, which we received on previous step. To get a sticker pack we should call `getPackData(sticker-pack-id)`, the return type is  `["bytes4[]" "address" "bool" "uint256" "uint256" "bytes"]` which represents the following fields: `[category owner mintable timestamp price contenthash]`. Price is the SNT value in wei setted by sticker pack owner. The contenthash is the IPFS address described in the [submit desction](#submit-a-sticker-pack) above. Other fields specification could be found in [Sticker Market spec](https://github.com/status-im/sticker-market/blob/651e88e5f38c690e57ecaad47f46b9641b8b1e27/docs/specification.md)
```
Comment (flexsurfer) [inline docs/draft/sticker-pack.md]: submit desction ?

Comment (Ferossgp) [inline docs/draft/sticker-pack.md]:
```suggestion
ID's are represented as `uint256` and are incremental from `0` to total number of sticker packs in contract, which we received on previous step. To get a sticker pack we should call `getPackData(sticker-pack-id)`, the return type is  `["bytes4[]" "address" "bool" "uint256" "uint256" "bytes"]` which represents the following fields: `[category owner mintable timestamp price contenthash]`. Price is the SNT value in wei setted by sticker pack owner. The contenthash is the IPFS address described in the [submit description](#submit-a-sticker-pack) above. Other fields specification could be found in [Sticker Market spec](https://github.com/status-im/sticker-market/blob/651e88e5f38c690e57ecaad47f46b9641b8b1e27/docs/specification.md)
```

**Review by decanus — APPROVED** [review]

**Pull request merged**